File: sdks/py/src/jeffs_brain_memory/cli/commands/serve.py
# SPDX-License-Identifier: Apache-2.0
"""`memory serve` — start the HTTP daemon."""

from __future__ import annotations

import os

import click


@click.command()
@click.option(
    "--addr",
    default=None,
    help="Address to bind (host:port). Defaults to $JB_ADDR or :8080.",
)
@click.option(
    "--root",
    default=None,
    help="JB_HOME directory (default $JB_HOME or ~/.jeffs-brain).",
)
@click.option(
    "--auth-token",
    default=None,
    help="Shared bearer token (default $JB_AUTH_TOKEN, optional).",
)
@click.option(
    "--contextualise/--no-contextualise",
    default=None,
    help="Enable live extraction contextualisation.",
)
@click.option(
    "--contextualise-cache-dir",
    default=None,
    help="Optional cache directory for live extraction contextualisation.",
)
def serve(
    addr: str | None,
    root: str | None,
    auth_token: str | None,
    contextualise: bool | None,
    contextualise_cache_dir: str | None,
) -> None:
    """Start the HTTP daemon matching `spec/PROTOCOL.md`."""
    import uvicorn

    from ...http.server import create_app

    resolved_addr = addr or os.environ.get("JB_ADDR") or ":8080"
    host, port = _parse_addr(resolved_addr)
    token = auth_token or os.environ.get("JB_AUTH_TOKEN")
    resolved_contextualise = (
        contextualise
        if contextualise is not None
        else _env_enabled(os.environ.get("JB_CONTEXTUALISE"))
    )
    app = create_app(
        root=root,
        auth_token=token,
        contextualise=resolved_contextualise,
        contextualise_cache_dir=(
            contextualise_cache_dir
            or os.environ.get("JB_CONTEXTUALISE_CACHE_DIR")
        ),
    )
    uvicorn.run(app, host=host, port=port, log_level="info")


def _parse_addr(value: str) -> tuple[str, int]:
    """Parse a host:port string into a (host, port) tuple.

    Leading colon (`:8080`) binds all interfaces defaulted to 127.0.0.1
    for safety. IPv6 literals with brackets are tolerated.
    """
    host_part, _, port_part = value.rpartition(":")
    if not port_part:
        raise click.BadParameter(f"--addr missing port: {value!r}")
    host = host_part.strip()
    if host.startswith("["):
        host = host[1:]
    if host.endswith("]"):
        host = host[:-1]
    if not host:
        host = "127.0.0.1"
    try:
        port = int(port_part)
    except ValueError as exc:
        raise click.BadParameter(f"--addr port must be numeric: {value!r}") from exc
    return host, port


def _env_enabled(value: str | None) -> bool:
    if value is None:
        return False
    return value.strip().lower() in {"1", "true", "yes", "on"}

File: sdks/go/cmd/memory/serve.go
// SPDX-License-Identifier: Apache-2.0

package main

import (
	"context"
	"log/slog"
	"net/http"
	"os"
	"os/signal"
	"strings"
	"syscall"

	"github.com/spf13/cobra"

	"github.com/jeffs-brain/memory/go/internal/httpd"
)

// Environment variable names recognised by the daemon. Mirror the
// JB_ namespace from the rollout plan so callers can feed identical
// env vars to both this binary and the TS SDK.
const (
	envHome                  = "JB_HOME"
	envAddr                  = "JB_ADDR"
	envAuthToken             = "JB_AUTH_TOKEN"
	envContextualise         = "JB_CONTEXTUALISE"
	envContextualiseCacheDir = "JB_CONTEXTUALISE_CACHE_DIR"
)

func serveCmd() *cobra.Command {
	var (
		addr               string
		root               string
		token              string
		contextualise      bool
		contextualiseCache string
	)
	cmd := &cobra.Command{
		Use:   "serve",
		Short: "Run the memory HTTP daemon",
		Long: "serve starts the HTTP daemon that speaks the Jeffs Brain " +
			"wire protocol documented at spec/PROTOCOL.md.",
		RunE: func(cmd *cobra.Command, args []string) error {
			if root == "" {
				root = os.Getenv(envHome)
			}
			if token == "" {
				token = os.Getenv(envAuthToken)
			}
			if cmd.Flags().Changed("contextualise") {
				if contextualise {
					_ = os.Setenv(envContextualise, "1")
				} else {
					_ = os.Unsetenv(envContextualise)
				}
			}
			if cmd.Flags().Changed("contextualise-cache-dir") {
				if strings.TrimSpace(contextualiseCache) == "" {
					_ = os.Unsetenv(envContextualiseCacheDir)
				} else {
					_ = os.Setenv(envContextualiseCacheDir, contextualiseCache)
				}
			}
			log := slog.New(slog.NewJSONHandler(os.Stdout, nil))

			ctx, cancel := signal.NotifyContext(cmd.Context(), os.Interrupt, syscall.SIGTERM)
			defer cancel()

			daemon, err := NewDaemon(ctx, root, token, log)
			if err != nil {
				return err
			}
			defer func() { _ = daemon.Close() }()

			srv := httpd.NewServer(addr, log)
			if token != "" {
				srv.SetAuthToken(token)
			}
			srv.AddRouteRegistrar(daemon.RegisterRoutes)
			srv.AddShutdownHook(func(_ context.Context) error {
				return daemon.Close()
			})

			return srv.Run(ctx)
		},
	}
	defaultAddr := os.Getenv(envAddr)
	if defaultAddr == "" {
		defaultAddr = ":8080"
	}
	cmd.Flags().StringVar(&addr, "addr", defaultAddr, "address to bind (host:port)")
	cmd.Flags().StringVar(&root, "root", "", "JB_HOME directory (default $JB_HOME or ~/.jeffs-brain)")
	cmd.Flags().StringVar(&token, "auth-token", "", "shared bearer token (default $JB_AUTH_TOKEN, optional)")
	cmd.Flags().BoolVar(&contextualise, "contextualise", false, "Enable live extraction contextualisation so extracted facts carry a situating prefix.")
	cmd.Flags().StringVar(&contextualiseCache, "contextualise-cache-dir", "", "Optional cache directory for live extraction contextualisation.")
	return cmd
}

// RegisterRoutes wires every endpoint defined by the daemon onto mux.
// The shape mirrors spec/PROTOCOL.md plus the brain-management endpoints
// in spec/MCP-TOOLS.md.
func (d *Daemon) RegisterRoutes(mux *http.ServeMux) {
	mux.HandleFunc("GET /v1/brains", d.handleListBrains)
	mux.HandleFunc("POST /v1/brains", d.handleCreateBrain)
	mux.HandleFunc("GET /v1/brains/{brainId}", d.handleGetBrain)
	mux.HandleFunc("DELETE /v1/brains/{brainId}", d.handleDeleteBrain)

	mux.HandleFunc("GET /v1/brains/{brainId}/documents/read", d.handleDocRead)
	mux.HandleFunc("GET /v1/brains/{brainId}/documents/stat", d.handleDocStat)
	mux.HandleFunc("GET /v1/brains/{brainId}/documents", d.handleDocListOrHead)
	mux.HandleFunc("HEAD /v1/brains/{brainId}/documents", d.handleDocHead)
	mux.HandleFunc("PUT /v1/brains/{brainId}/documents", d.handleDocPut)
	mux.HandleFunc("POST /v1/brains/{brainId}/documents/append", d.handleDocAppend)
	mux.HandleFunc("DELETE /v1/brains/{brainId}/documents", d.handleDocDelete)
	mux.HandleFunc("POST /v1/brains/{brainId}/documents/rename", d.handleDocRename)
	mux.HandleFunc("POST /v1/brains/{brainId}/documents/batch-ops", d.handleDocBatch)

	mux.HandleFunc("POST /v1/brains/{brainId}/search", d.handleSearch)
	mux.HandleFunc("POST /v1/brains/{brainId}/ask", d.handleAsk)

	mux.HandleFunc("POST /v1/brains/{brainId}/ingest/file", d.handleIngestFile)
	mux.HandleFunc("POST /v1/brains/{brainId}/ingest/url", d.handleIngestURL)

	mux.HandleFunc("POST /v1/brains/{brainId}/remember", d.handleRemember)
	mux.HandleFunc("POST /v1/brains/{brainId}/recall", d.handleRecall)
	mux.HandleFunc("POST /v1/brains/{brainId}/extract", d.handleExtract)
	mux.HandleFunc("POST /v1/brains/{brainId}/reflect", d.handleReflect)
	mux.HandleFunc("POST /v1/brains/{brainId}/consolidate", d.handleConsolidate)

	mux.HandleFunc("GET /v1/brains/{brainId}/events", d.handleEvents)
}
